File: sandbox/plc4go/internal/plc4go/modbus/readwrite/model/ModbusPDUGetComEventCounterResponse.go
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
package model

import (
    "errors"
    "plc4x.apache.org/plc4go-modbus-driver/v0/internal/plc4go/spi"
    "plc4x.apache.org/plc4go-modbus-driver/v0/internal/plc4go/utils"
)

// The data-structure of this message
type ModbusPDUGetComEventCounterResponse struct {
    Status uint16
    EventCount uint16
    ModbusPDU
}

// The corresponding interface
type IModbusPDUGetComEventCounterResponse interface {
    IModbusPDU
    Serialize(io utils.WriteBuffer) error
}

// Accessors for discriminator values.
func (m ModbusPDUGetComEventCounterResponse) ErrorFlag() bool {
    return false
}

func (m ModbusPDUGetComEventCounterResponse) FunctionFlag() uint8 {
    return 0x0B
}

func (m ModbusPDUGetComEventCounterResponse) Response() bool {
    return true
}

func (m ModbusPDUGetComEventCounterResponse) initialize() spi.Message {
    return m
}

func NewModbusPDUGetComEventCounterResponse(status uint16, eventCount uint16) ModbusPDUInitializer {
    return &ModbusPDUGetComEventCounterResponse{Status: status, EventCount: eventCount}
}

func CastIModbusPDUGetComEventCounterResponse(structType interface{}) IModbusPDUGetComEventCounterResponse {
    castFunc := func(typ interface{}) IModbusPDUGetComEventCounterResponse {
        if iModbusPDUGetComEventCounterResponse, ok := typ.(IModbusPDUGetComEventCounterResponse); ok {
            return iModbusPDUGetComEventCounterResponse
        }
        return nil
    }
    return castFunc(structType)
}

func CastModbusPDUGetComEventCounterResponse(structType interface{}) ModbusPDUGetComEventCounterResponse {
    castFunc := func(typ interface{}) ModbusPDUGetComEventCounterResponse {
        if sModbusPDUGetComEventCounterResponse, ok := typ.(ModbusPDUGetComEventCounterResponse); ok {
            return sModbusPDUGetComEventCounterResponse
        }
        if sModbusPDUGetComEventCounterResponse, ok := typ.(*ModbusPDUGetComEventCounterResponse); ok {
            return *sModbusPDUGetComEventCounterResponse
        }
        return ModbusPDUGetComEventCounterResponse{}
    }
    return castFunc(structType)
}

func (m ModbusPDUGetComEventCounterResponse) LengthInBits() uint16 {
    var lengthInBits uint16 = m.ModbusPDU.LengthInBits()

    // Simple field (status)
    lengthInBits += 16

    // Simple field (eventCount)
    lengthInBits += 16

    return lengthInBits
}

func (m ModbusPDUGetComEventCounterResponse) LengthInBytes() uint16 {
    return m.LengthInBits() / 8
}

func ModbusPDUGetComEventCounterResponseParse(io *utils.ReadBuffer) (ModbusPDUInitializer, error) {

    // Simple Field (status)
    status, _statusErr := io.ReadUint16(16)
    if _statusErr != nil {
        return nil, errors.New("Error parsing 'status' field " + _statusErr.Error())
    }

    // Simple Field (eventCount)
    eventCount, _eventCountErr := io.ReadUint16(16)
    if _eventCountErr != nil {
        return nil, errors.New("Error parsing 'eventCount' field " + _eventCountErr.Error())
    }

    // Create the instance
    return NewModbusPDUGetComEventCounterResponse(status, eventCount), nil
}

func (m ModbusPDUGetComEventCounterResponse) Serialize(io utils.WriteBuffer) error {
    ser := func() error {

    // Simple Field (status)
    status := uint16(m.Status)
    _statusErr := io.WriteUint16(16, (status))
    if _statusErr != nil {
        return errors.New("Error serializing 'status' field " + _statusErr.Error())
    }

    // Simple Field (eventCount)
    eventCount := uint16(m.EventCount)
    _eventCountErr := io.WriteUint16(16, (eventCount))
    if _eventCountErr != nil {
        return errors.New("Error serializing 'eventCount' field " + _eventCountErr.Error())
    }

        return nil
    }
    return ModbusPDUSerialize(io, m.ModbusPDU, CastIModbusPDU(m), ser)
}

File: plc4py/plc4py/protocols/modbus/readwrite/ModbusPDUGetComEventCounterResponse.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from dataclasses import dataclass

from plc4py.api.exceptions.exceptions import PlcRuntimeException
from plc4py.api.exceptions.exceptions import SerializationException
from plc4py.api.messages.PlcMessage import PlcMessage
from plc4py.protocols.modbus.readwrite.ModbusPDU import ModbusPDU
from plc4py.spi.generation.ReadBuffer import ReadBuffer
from plc4py.spi.generation.WriteBuffer import WriteBuffer
from plc4py.utils.ConnectionStringHandling import strtobool
from typing import ClassVar
import math


@dataclass
class ModbusPDUGetComEventCounterResponse(ModbusPDU):
    status: int
    event_count: int
    # Accessors for discriminator values.
    error_flag: ClassVar[bool] = False
    function_flag: ClassVar[int] = 0x0B
    response: ClassVar[bool] = True

    def serialize_modbus_pdu_child(self, write_buffer: WriteBuffer):
        write_buffer.push_context("ModbusPDUGetComEventCounterResponse")

        # Simple Field (status)
        write_buffer.write_unsigned_short(
            self.status, bit_length=16, logical_name="status"
        )

        # Simple Field (eventCount)
        write_buffer.write_unsigned_short(
            self.event_count, bit_length=16, logical_name="eventCount"
        )

        write_buffer.pop_context("ModbusPDUGetComEventCounterResponse")

    def length_in_bytes(self) -> int:
        return int(math.ceil(float(self.length_in_bits() / 8.0)))

    def length_in_bits(self) -> int:
        length_in_bits: int = super().length_in_bits()
        _value: ModbusPDUGetComEventCounterResponse = self

        # Simple field (status)
        length_in_bits += 16

        # Simple field (eventCount)
        length_in_bits += 16

        return length_in_bits

    @staticmethod
    def static_parse_builder(read_buffer: ReadBuffer, response: bool):
        read_buffer.push_context("ModbusPDUGetComEventCounterResponse")

        if isinstance(response, str):
            response = bool(strtobool(response))

        status: int = read_buffer.read_unsigned_short(
            logical_name="status", bit_length=16, response=response
        )

        event_count: int = read_buffer.read_unsigned_short(
            logical_name="event_count", bit_length=16, response=response
        )

        read_buffer.pop_context("ModbusPDUGetComEventCounterResponse")
        # Create the instance
        return ModbusPDUGetComEventCounterResponseBuilder(status, event_count)

    def equals(self, o: object) -> bool:
        if self == o:
            return True

        if not isinstance(o, ModbusPDUGetComEventCounterResponse):
            return False

        that: ModbusPDUGetComEventCounterResponse = ModbusPDUGetComEventCounterResponse(
            o
        )
        return (
            (self.status == that.status)
            and (self.event_count == that.event_count)
            and super().equals(that)
            and True
        )

    def hash_code(self) -> int:
        return hash(self)

    def __str__(self) -> str:
        # TODO:- Implement a generic python object to probably json convertor or something.
        return ""


@dataclass
class ModbusPDUGetComEventCounterResponseBuilder:
    status: int
    event_count: int

    def build(
        self,
    ) -> ModbusPDUGetComEventCounterResponse:
        modbus_pduget_com_event_counter_response: (
            ModbusPDUGetComEventCounterResponse
        ) = ModbusPDUGetComEventCounterResponse(self.status, self.event_count)
        return modbus_pduget_com_event_counter_response
